ci/docker/vmcontrol.py

```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/env python3
+
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# -*- coding: utf-8 -*-
+"""Utilities to control a guest VM"""
+
+__author__ = 'Pedro Larroy'
+__version__ = '0.1'
+
+import os
+import sys
+import subprocess
+import argparse
+import logging
+from subprocess import call, check_call, Popen, DEVNULL, PIPE
+import time
+import sys
+import multiprocessing
+import shlex
+
+###################################################
+#
+# Virtual testing with QEMU
+#
+# We start QEMU instances that have a local port in the host redirected to the ssh port.
+#
+# The VMs are provisioned after boot, tests are run and then they are stopped
+#
+
+QEMU_RUN="""
+qemu-system-arm -M virt -m {ram} \
+  -kernel vmlinuz \
+  -initrd initrd.img \
+  -append 'root=/dev/vda1' \
+  -drive if=none,file=vda.qcow2,format=qcow2,id=hd \
+  -device virtio-blk-device,drive=hd \
+  -netdev user,id=mynet,hostfwd=tcp::{ssh_port}-:22 \
+  -device virtio-net-device,netdev=mynet \
+  -display none -nographic
+"""
+
+class VMError(RuntimeError):
+    pass
+
+class VM:
+    """Control of the virtual machine"""
+    def __init__(self, ssh_port=2222):
+        self.log = logging.getLogger(VM.__name__)
+        self.ssh_port = ssh_port
+        self.timeout_s = 300
+        #self.timeout_s = None
+        self.qemu_process = None
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.terminate()
+
+    def start(self):
+        self.log.info("Starting VM, ssh port redirected to localhost:%s", self.ssh_port)
+        if self.running():
+            raise VMError("VM is running, shutdown first")
+        self.qemu_process = run_qemu(self.ssh_port)
+        # give a bit of time for early execution failures
+        time.sleep(0)
```

Comment:
That's returning as soon as the scheduler comes back to the process, so that's not very long. Can we maybe have a better way to determine files and maybe wait for a specific event?

Reply:
I removed this code, is not required anymore.